src/hotspot/cpu/aarch64/vm_version_aarch64.cpp

```diff
@@ -259,6 +259,9 @@ void VM_Version::initialize() {
     if (FLAG_IS_DEFAULT(UseCryptoPmullForCRC32)) {
       FLAG_SET_DEFAULT(UseCryptoPmullForCRC32, true);
     }
+    if (FLAG_IS_DEFAULT(CodeEntryAlignment)) {
+      FLAG_SET_DEFAULT(CodeEntryAlignment, 32);
```

Comment:
Why only for V1/V2? What about N1/N2?

Reply:
I’d like to see the same for N1.

Reply:
There are reasons against the change for N1/N2
- Vladimir saw a regression when testing -XX:CodeEntryAlignment=16 on Ampere (Neoverse N1)
- with native codes experiment I found that code entry 64B is preferable on G2 (Neoverse N1)

Reply:
> There are reasons against the change for N1/N2
> 
>     * Vladimir saw a regression when testing -XX:CodeEntryAlignment=16 on Ampere (Neoverse N1)

The regression is for 16 not 32. Maybe 32 won't cause regressions. 

> 
>     * with native codes experiment I found that code entry 64B is preferable on G2 (Neoverse N1)

Could you share more data?

Reply:
> Could you share more data?

This is a simple native benchmark: https://cr.openjdk.org/~bulasevich/AlignBench.cpp

The program tests the performance of code starting from different alignments 
using a generated block of code that primarily consists of NOP (no operation)
and RET (return from subroutine) instructions.

In this experiment (though it may not be meaningful for real-world scenarios) 
on G2 execution time increases with misalignment from the 64-byte boundary, with the 
best performance at exact 64-byte alignment and slower times as the offset grows.

align_64+0: 391.457ms 391.157ms 391.481ms
align_64+4: 392.245ms 392.057ms 391.792ms
align_64+8: 392.154ms 392.962ms 392.65ms
align_64+12: 393.485ms 393.307ms 393.658ms
align_64+16: 394.434ms 394.679ms 394.284ms
align_64+20: 395.681ms 395.709ms 394.843ms
align_64+24: 395.799ms 396.396ms 395.977ms
align_64+28: 397.379ms 397.278ms 397.359ms
align_64+32: 397.677ms 397.82ms 397.95ms
align_64+36: 399.08ms 398.88ms 399.075ms
align_64+40: 399.829ms 400.118ms 399.981ms
align_64+44: 400.916ms 400.747ms 401.241ms
align_64+48: 401.736ms 401.831ms 402.54ms
align_64+52: 402.705ms 402.569ms 402.446ms
align_64+56: 403.718ms 403.822ms 403.535ms
align_64+60: 404.722ms 404.824ms 404.726ms
align_64+64: 390.852ms 390.669ms 391.051ms

Reply:
This benchmark certainly has some limitations as nops are special and can be discarded and having the 64B alignment means the core can quickly fetch and throw away as many nops as possible. I'm not sure that it's particularly representative of a real output and the benefits that code density can bring here.

Reply:
The behavior remains unchanged when replacing the NOP with an ADD x1,x1,x1 instruction. That said, I fully agree with you that the benchmark is peculiar, and the result doesn't necessarily indicate whether the platform is sensitive to code entry alignment. Additionally, I'd like to point out that on the same N1 platform, EEMBC's CoreMark benchmark runs 0.07% faster (a small difference, I know) on G2 when built with -falign-functions=64 compared to -falign-functions=16, with the result for -falign-functions=32 falling in between. This makes me doubt that CodeEntryAlignment=16/32 is reasonable for N1.

Let me ask Vladimir if it is possible to check the performance with CodeEntryAlignment=32 setting.

Reply:
Dear @vnkozlov,
Sorry for bothering you. Is it possible to check if CodeEntryAlignment=32 setting causes the regression on Ampere system with the internal benchmark?

Reply:
Okay

Reply:
We pushed into OpenJDK JMH benchmark which is similar to what we use internally:
test/micro/org/openjdk/bench/vm/compiler/CodeCacheStress.java

You can try it while we are running our testing.

Reply:
CodeEntryAlignment=32 has the same regression as CodeEntryAlignment=12.
And CodeEntryAlignment=64 has no regression

Reply:
All runs with CodeCacheSegmentSize=64

Reply:
Thank you!

Reply:
The public CodeCacheStress benchmark is very noisy, easily giving +-10% on subsequent runs. On longer runs, CodeEntryAlignment=16 give ~1% regression on Neoverse N1 (Graviton2 and Ampere Altra), but CodeEntryAlignment=32 is OK. My measurements:
- Graviton2 (Neoverse N1) results:
    - CodeEntryAlignment=16: 93.33 ± 0.60 us/op
    - CodeEntryAlignment=32: 92.10 ± 0.69 us/op
    - CodeEntryAlignment=64: 92.67 ± 0.62 us/op
- Ampere Altra (Neoverse N1) results:
    - CodeEntryAlignment=16: 85.50 ± 0.88 us/op
    - CodeEntryAlignment=32: 83.79 ± 0.97 us/op
    - CodeEntryAlignment=64: 84.10 ± 0.82 us/op
- Graviton4 (Neoverse V2) results:
    - CodeEntryAlignment=16: 34.60 ± 0.41 us/op
    - CodeEntryAlignment=32: 34.30 ± 0.35 us/op
    - CodeEntryAlignment=64: 34.38 ± 0.40 us/op